### DISCUSSION TITLE
Add Typescript types for all app-router component file conventions

### DISCUSSION THREAD

**Discussion by kpervin** [vercel/next.js#65801] [Ideas]
### Goals

Add Typescript types for new app-router file conventions.

### Background

Previously for the pages router we had type exports for `getServerSideProps`, `NextPage`, etc. For the new file conventions, having typescript types for the functions would be ideal.

### Proposal

For the new app router, having types for `layout.tsx`, `page.tsx`, etc would be optimal as we would be able to know about using:

- layout props: `{ children: ReactNode; params?: Record<string, string> }`
- page props: `{ params?: Record<string, string>; searchParams?: Record<string, string> }`

and so on. Having these in typescript types reduces the need to go to the documentation.

**Comment by Netail** [comment]
As we now also have generated types for next/link (typedRoutes), it would also be nice to automatically have params generated in these types
